### PR TITLE
Save details to firebase

### PR DESCRIPTION
**Pull request by peytonshetler**
## Summary
-Created an AddNewLocationViewController where users enter info about a location (name, city, state, category) and it's saved to Firebase.  This doesn't include testing because I wasn't really sure what to test since I already have guard statements guarding against nil values in the textFields (but I'm sure there's something haha).

#### Characteristics

- [X] Feature completed
- [X] Experimental
- [ ] UI affected (screen shots included)

### CODE REVIEW

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        // guarding against nil values
+        guard let name = nameTextField.text, let city = cityTextField.text,
+            let state = stateTextField.text, let category = categoryTextField.text else {
+            print("Location inputs are nil")
+            return
+        }
+        // capture and structure the new location
+        // Valid data types for Firebase are String, Number, Array and
+        // Dictionary.  Not sure if custom Location type will work(?)
+        let newLocation = [
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/AddNewLocationViewController.swift]: Because location is of type `Codable` you could literally send a coded `location` directly to Firebase without having to create a dictionary. [This link](https://stackoverflow.com/a/46329055/9333764) should point you in the right direction.

File: Been-There/Been-There/Controller/HomeScreenViewController.swift [hunk 2]
```diff
@@ -6,13 +6,16 @@
...
 
-class HomeScreenViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
+class HomeScreenViewController: UIViewController, UICollectionViewDelegate,
+        UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
     
     let cellId = "cellId"
     let topContainer = UIView()
     let bottomContainer = UIView()
     let screenTitle = UILabel()
+    let collectionArray = ["New Location", "My Locations", "Item 3"]
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/HomeScreenViewController.swift]: What is Item3?

File: Been-There/Been-There/Client/Alert/BasicAlert.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+//  Created by Peyton Shetler on 4/30/19.
+//  Copyright © 2019 Peyton Shetler. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+struct Alert {
+    
+    static func showBasicAlert(on viewController: UIViewController,
```
[the remark this answers is not in the document]
Comment (peytonshetler) [reply Been-There/Been-There/Client/Alert/BasicAlert.swift]: Thanks!!

File: Been-There/Been-There/Controller/HomeScreenViewController.swift (hunk 2 above)
Comment (peytonshetler) [reply Been-There/Been-There/Controller/HomeScreenViewController.swift]: I've re-named it, it's basically just a placeholder.

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = UIColor.forestGreen()
+        // Do any additional setup after loading the view.
+    }
+    
+    func saveDataToFirebase() {
+        // firebase database reference
+        let ref = Database.database().reference()
```
[the remark this answers is not in the document]
Comment (peytonshetler) [reply Been-There/Been-There/Controller/AddNewLocationViewController.swift]: I believe we solved this on our call

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift (hunk 1 above)
Comment (peytonshetler) [reply Been-There/Been-There/Controller/AddNewLocationViewController.swift]: We delved into this as well on our call

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift [hunk 5]
```diff
@@ -73,4 +57,15 @@ class AddNewLocationViewController: UIViewController {
...
+    // Objective overview:  The "Add" button will start a timer that will slowly increment.  Upon completion of the "saveLocationToFirebase" func, the progress bar will be updated to a value of 1 (completion).
+    
+    //  Steps
+    // 1. Progress view style: give it rounded corners
+    
+    // 2. Progress view functionality:
+    //      -instantiate a timer.  ex: let timer - Timer()
+    //      -create an updateProgress function to update the progressBar status
+    //      -create a scheduledTimer (timer.scheduledTimer) function that will call the updateProgress func
 }
```
Comment (peytonshetler) [inline Been-There/Been-There/Controller/AddNewLocationViewController.swift]: These are my plans for the the progress bar on the New Location View Controller.

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift [hunk 6]
```diff
@@ -17,53 +17,37 @@ class AddNewLocationViewController: UIViewController {
...
-        saveDataToFirebase()
+        firebase.saveLocationToFirebase(Location(name: nameTextField.text,
+            city: cityTextField.text, state: stateTextField.text,
+            category: categoryTextField.text,
+            lat: nil, long: nil)) { [weak self] (error) in
+                if let error = error {
+                    if self != nil {
+                        Alert.showBasicAlert(on: self!, withTitle: "Uh-oh",
+                                             message: "Something went very wrong")
+                        print(error.localizedDescription)
```
Comment (peytonshetler) [inline Been-There/Been-There/Controller/AddNewLocationViewController.swift]: I've tested a few times without wifi and this alert isn't triggering.

File: Been-There/Been-There.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -36,6 +40,9 @@
 		79C0D886224D394700A5052A /* HomeScreenCollectionViewCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79C0D885224D394700A5052A /* HomeScreenCollectionViewCell.swift */; };
 		79C0D887224D394700A5052A /* HomeScreenCollectionViewCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79C0D885224D394700A5052A /* HomeScreenCollectionViewCell.swift */; };
 		79C0D888224D394700A5052A /* HomeScreenCollectionViewCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79C0D885224D394700A5052A /* HomeScreenCollectionViewCell.swift */; };
+		79C87F66227A7D2700620EA9 /* encodeToJson.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79C87F65227A7D2700620EA9 /* encodeToJson.swift */; };
+		79C87F67227A7D2700620EA9 /* encodeToJson.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79C87F65227A7D2700620EA9 /* encodeToJson.swift */; };
+		79C87F68227A7D2700620EA9 /* encodeToJson.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79C87F65227A7D2700620EA9 /* encodeToJson.swift */; };
```
Comment (peytonshetler) [inline Been-There/Been-There.xcodeproj/project.pbxproj]: I moved the encodeToJson function from the Extension group to a new group ( Global -> Functions -> encodeToJson.swift ) since it wasn't even an extension, haha

File: Been-There/Been-There.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -228,22 +224,6 @@
...
-			sourceTree = "<group>";
-		};
-		79706C942280DBFE00A8C675 /* Functions */ = {
-			isa = PBXGroup;
-			children = (
-				79C87F65227A7D2700620EA9 /* encodeToJson.swift */,
-			);
-			path = Functions;
-			sourceTree = "<group>";
-		};
```
Comment (peytonshetler) [inline Been-There/Been-There.xcodeproj/project.pbxproj]: Deleted the encodeToJson function as we mentioned earlier 👍🏻

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift (hunk 6 above)
Comment (JZDesign) [reply Been-There/Been-There/Controller/AddNewLocationViewController.swift]: That's because I was wrong. The completion block never fires in the proxy while the network is down.

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift (hunk 5 above)
Comment (JZDesign) [reply Been-There/Been-There/Controller/AddNewLocationViewController.swift]: I would use Apple's [Activity Indicator](https://developer.apple.com/documentation/uikit/uiactivityindicatorview) You could literally define one in the View, add it on `viewDidLoad` then `show` it when you press the button and `stop` it on completion.

File: Been-There/Been-There/Global/Functions/encodeToJson.swift [hunk 9]
```diff
@@ -0,0 +1,30 @@
+//
```
Comment (JZDesign) [inline Been-There/Been-There/Global/Functions/encodeToJson.swift]: You can delete this file before you merge

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift (hunk 6 above)
Comment (JZDesign) [reply Been-There/Been-There/Controller/AddNewLocationViewController.swift]: This should still work though if Firebase returns an error.

File: Been-There/Been-There/Controller/AddNewLocationViewController.swift (hunk 5 above)
Comment (peytonshetler) [reply Been-There/Been-There/Controller/AddNewLocationViewController.swift]: Will do!
